File: backend-career/internshiproutes.js
/**
 * internshipRoutes.js
 * ---------------------------------------------------------
 * Example Express routes showing how adzunaIntegration.js
 * plugs into your existing server. Mount this in your
 * main server.js with:
 *
 *   const internshipRoutes = require('./internshipRoutes');
 *   app.use('/api/internships', internshipRoutes);
 *
 * Also call startScheduledSync() once, on server startup,
 * so the Firestore cache stays fresh automatically.
 * ---------------------------------------------------------
 */

const express = require('express');
const router = express.Router();
const {
  getMatchedInternshipsForStudent,
  syncAllInternships
} = require('./adzunaintegration');

/**
 * GET /api/internships?roleType=frontend
 * Requires the logged-in student's skills — normally you'd
 * pull this from their Firestore user profile via req.user.uid
 * (from your existing auth middleware), not from the query string.
 * Shown here simplified for clarity.
 */
router.get('/', async (req, res) => {
  try {
    const { roleType, skills } = req.query;

    // In your real app: fetch studentSkills from Firestore using
    // the authenticated user's UID instead of trusting the query param.
    const studentSkills = skills ? skills.split(',') : [];

    const matches = await getMatchedInternshipsForStudent(studentSkills, roleType);
    res.json({ count: matches.length, results: matches });
  } catch (err) {
    console.error('Failed to fetch matched internships:', err);
    res.status(500).json({ error: 'Could not load internships right now.' });
  }
});

/**
 * POST /api/internships/sync
 * Manual trigger for testing — lets you refresh the Firestore
 * cache on demand instead of waiting for the 6-hour cron job.
 * In production, protect this route (admin-only) so it can't
 * be spammed and burn through your Adzuna rate limit.
 *
 * NOTE: the `detail` field in the error response is for local
 * debugging only. Remove it (or gate it behind an env check)
 * before deploying, so you don't leak internal error messages
 * to the public.
 */
router.post('/sync', async (req, res) => {
  try {
    await syncAllInternships();
    res.json({ message: 'Sync complete' });
  } catch (err) {
    console.error('SYNC ERROR:', err); // full object -> prints stack trace in Terminal 1
    res.status(500).json({
      error: 'Sync failed',
      detail: err.message // temporary — remove before production
    });
  }
});

module.exports = router;

File: backend/internship_routes.py
from flask import Blueprint, request, jsonify
from adzuna_integration import (
    get_matched_internships_for_student,
    sync_all_internships
)

internship_bp = Blueprint("internship_bp", __name__)


@internship_bp.route("", methods=["GET"])
def get_internships():
    try:
        role_type = request.args.get("roleType")
        skills = request.args.get("skills")
        student_skills = skills.split(",") if skills else []
        matches = get_matched_internships_for_student(student_skills, role_type)
        return jsonify({"count": len(matches), "results": matches})
    except Exception as e:
        print("Failed to fetch matched internships:", str(e))
        return jsonify({"error": "Could not load internships right now."}), 500


@internship_bp.route("/sync", methods=["POST"])
def sync():
    try:
        sync_all_internships()
        return jsonify({"message": "Sync complete"})
    except Exception as e:
        print("SYNC ERROR:", str(e))
        return jsonify({"error": "Sync failed"}), 500